lib/improver/utilities/indexing_operations.py

```diff
@@ -0,0 +1,168 @@
+            sub-arrays from which values are to be extracted at positions that
+            match those given in the index_array.
+    Returns:
+        result (np.array):
+            An array containing data extracted from the array_set. The returned
+            array will have the same shape as the sub-arrays unless the
+            index_array has dimensions equivalent to array_set. In this case
+            the returned array will have an equivalent number of dimensions
+            (i.e. one more than the sub-arrays).
+    Raises:
+        ValueError: If input arrays cannot be broadcast to a common shape.
+        IndexError: If dimensions are added to array_set during the
+                    broadcasting of arrays; this will prevent the later
+                    indexing from working as expected.
+        IndexError: If an index exceeds the number of available sub-arrays.
+    """
+    # Used to correct array shape in output following intermediate broadcast.
+    index_array = np.array(index_array)
+    index_dims = index_array.ndim
+    trim_result = index_array.shape != array_set.shape
+
+    # Broadcast arrays to matching shape. This is included to replicate as much
+    # of the numpy choose functionality as possible.
+    try:
+        broadcast_arrays = np.broadcast_arrays(index_array, array_set)
+    except ValueError as err:
+        msg = '{}\nindex_array shape: {}\narray_set shape: {}'.format(
+            err, index_array.shape, array_set.shape)
+        raise ValueError(msg)
+
+    # numpy choose can handle the condition below, but we currently have no
+    # need to accommodate this behaviour, so we trap it and raise an error.
+    if broadcast_arrays[1].ndim > array_set.ndim:
+        msg = ('Dimensionality of array_set has increased which will prevent '
+               'indexing from working as expected.')
+        raise IndexError(msg)
+
+    index_array = broadcast_arrays[0]
+    array_set = broadcast_arrays[1]
+
+    # Use indexing to extract the requires values from the sub-arrays.
+    try:
+        result = np.array(
+            [array_set[index_array[I]][I[1:]]
+             for I in np.lib.index_tricks.ndindex(index_array.shape)]
+        ).reshape(index_array.shape)
+    except IndexError as err:
+        msg = ('{}\nindex_array contains an index that is larger than the '
+               'number of sub-arrays from which data can be drawn.\nMax '
+               'index: {}\nNumber of sub-arrays: {}'.format(
+                   err, index_array.max(), array_set.shape[0]))
+        raise IndexError(msg)
+
+    # If the index_array shape did not match the array_set shape, it is assumed
+    # that the indexing is for the sub-arrays, not the complete array_set. The
+    # leading dimension is thus cut away in the result.
+    if trim_result:
+        result = result[0]
+
+    # Add back any lost empty dimensions to ensure the returned array has the
```

Comment:
This is a bit unclear - suggest `Add back any lost scalar / single point dimensions...`?

Reply:
Removed.